Ignore instance times from endElement*() for inactive timed elements

Ahmad-S792 opened the pull request:
#### a937e1c8a1a3da0a2f63dd3a1f22c59bcea4b5ad
<pre>
Ignore instance times from endElement*() for inactive timed elements

<a href="https://bugs.webkit.org/show_bug.cgi?id=200259">https://bugs.webkit.org/show_bug.cgi?id=200259</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: <a href="https://github.com/chromium/chromium/commit/810144f1502198bb66cd265a2ce9a16ccd106dd2">https://github.com/chromium/chromium/commit/810144f1502198bb66cd265a2ce9a16ccd106dd2</a>

When endElementAt()/endElement() is trying to add a new instance time
and there&apos;s no active interval, just ignore the new instance time.

Spec [1]:

   &quot;While the element is not active, any end specification of the
   event is ignored.&quot;

[1] <a href="https://www.w3.org/TR/2001/REC-smil-animation-20010904/#Timing-EventSensitivity">https://www.w3.org/TR/2001/REC-smil-animation-20010904/#Timing-EventSensitivity</a>

* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::endElementAt):
* Source/WebCore/svg/animation/SVGSMILElement.h:
(WebCore::SVGSMILElement::activeState const):
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/scripted/end-element-on-inactive-element-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/280184@main">https://commits.webkit.org/280184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57e563652ab3884316e5bfb7c21ed5b7403a2ef8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55934 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58933 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6366 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6562 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45042 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4399 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57963 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33151 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48228 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26180 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29934 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5556 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4509 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5827 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60520 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5947 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52470 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48297 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51987 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12398 "Found 1 new test failure: imported/w3c/web-platform-tests/webrtc/simulcast/getStats.https.html (failure)") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8269 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31087 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32171 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33253 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31919 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -756,8 +756,12 @@ void SVGSMILElement::addInstanceTime(BeginOrEnd beginOrEnd, SMILTime time, SMILT
+        return;
     auto& list = beginOrEnd == Begin ? m_beginTimes : m_endTimes;
-    list.append(SMILTimeWithOrigin(time, origin));
+    list.append(timeWithOrigin);
```
I do not think this is the right place to add this code. Since `SVGAnimationElement::endElementAt()` is the caller to `addInstanceTime()` which passes `End` and `SMILTimeWithOrigin::ScriptOrigin`, I think you can do the following:

In the `protected` section of `SVGSMILElement`, add this inline function:

```
ActiveState activeState() const { return m_activeState; }
```

And in `SVGAnimationElement::endElementAt()` add this if statement:

```
if (activeState() == Inactive)
    return;
addInstanceTime(End, elapsed() + offset, SMILTimeWithOrigin::ScriptOrigin);
```

Ahmad-S792 replied:
@shallawa - I get compile errors with suggested approach. :-( Will continue to explore.

shallawa replied:
I think you need to move the definition of `ActiveState` from the `private:` section to the `protected:` section:

```
protected:
    enum ActiveState { Inactive, Active, Frozen };
    ActiveState activeState() const { return m_activeState; }
    void setInactive() { m_activeState = Inactive; }
```

Ahmad-S792 replied:
Done in https://github.com/WebKit/WebKit/commit/0e646703be2b2c1f7bf2a8e6f34edf514887e5a1. Thanks for all the help @shallawa .

shallawa commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -257,6 +257,9 @@ void SVGAnimationElement::beginElementAt(float offset)
 void SVGAnimationElement::endElementAt(float offset)
 {
     ASSERT(std::isfinite(offset));
+    // Ignore new instance times for 'end' if the element is not active and the origin is script.
```
I think this comment does not add much. The code below is self explanatory.

Ahmad-S792 replied:
Removed in [57e5636](https://github.com/WebKit/WebKit/commit/57e563652ab3884316e5bfb7c21ed5b7403a2ef8).

shallawa commented on Source/WebCore/svg/animation/SVGSMILElement.cpp:
```diff
@@ -756,8 +756,9 @@ void SVGSMILElement::addInstanceTime(BeginOrEnd beginOrEnd, SMILTime time, SMILT
+    SMILTimeWithOrigin timeWithOrigin(time, origin);
     auto& list = beginOrEnd == Begin ? m_beginTimes : m_endTimes;
-    list.append(SMILTimeWithOrigin(time, origin));
+    list.append(timeWithOrigin);
```
I think creating an inline `SMILTimeWithOrigin` was better than storing it in a local variable and using it only once.

Ahmad-S792 replied:
I have plans for it in future merges. Although happy to revert.

Ahmad-S792 replied:
Done in [57e5636](https://github.com/WebKit/WebKit/commit/57e563652ab3884316e5bfb7c21ed5b7403a2ef8). Will reintroduce this whenever I need for future.

Commit:
Ignore instance times from endElement*() for inactive timed elements

https://bugs.webkit.org/show_bug.cgi?id=200259

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: https://github.com/chromium/chromium/commit/810144f1502198bb66cd265a2ce9a16ccd106dd2

When endElementAt()/endElement() is trying to add a new instance time
and there's no active interval, just ignore the new instance time.

Spec [1]:

   "While the element is not active, any end specification of the
   event is ignored."

[1] https://www.w3.org/TR/2001/REC-smil-animation-20010904/#Timing-EventSensitivity

* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::endElementAt):
* Source/WebCore/svg/animation/SVGSMILElement.h:
(WebCore::SVGSMILElement::activeState const):
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/scripted/end-element-on-inactive-element-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/280184@main
